pkg/kubelet/status/status_manager.go

```diff
@@ -676,20 +695,48 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 	}
 }
 
-// needsUpdate returns whether the status is stale for the given pod UID.
-// This method is not thread safe, and must only be accessed by the sync thread.
-func (m *manager) needsUpdate(uid types.UID, status versionedPodStatus) bool {
+// podForAPIServer returns the appropriate API pod (mirror or otherwise) for the given (kubelet) pod UID.
+func (m *manager) podForAPIServer(uid types.UID) (*v1.Pod, bool) {
+	pod, ok := m.podManager.GetPodByUID(types.UID(uid))
+	if !ok {
+		klog.V(4).InfoS("Pod has been deleted, no need to reconcile", "podUID", string(uid))
+		return nil, false
+	}
+	// If the pod is a static pod, we should check its mirror pod, because only status in mirror pod is meaningful to us.
+	if kubetypes.IsStaticPod(pod) {
+		mirrorPod, ok := m.podManager.GetMirrorPodByPod(pod)
+		if !ok {
+			klog.V(4).Infof("Static pod %q has no corresponding mirror pod, no need to reconcile", format.Pod(pod))
+			return nil, false
+		}
+		pod = mirrorPod
+	}
+	return pod, true
+}
+
+// rememberAPIStatus records the last version of the pod that we wrote to the API server.
+func (m *manager) rememberAPIStatus(uid kubetypes.MirrorPodUID, version uint64) {
+	m.podStatusesLock.Lock()
+	defer m.podStatusesLock.Unlock()
+
+	m.apiStatusVersions[uid] = version
+}
+
+// isStatusOutdated is true if the pod status is newer than what was last written
+// to the API server.
+func (m *manager) isStatusOutdated(uid types.UID, status versionedPodStatus) bool {
+	m.podStatusesLock.RLock()
```

Comment:
This method holds the read lock so it can check apiStatusVersions in case we parallelize this method.